### PR TITLE
feat: add configuration options for Goma

### PR DESCRIPTION
**Pull request by codebytere**
This PR adds configuration options for use with the Goma distibuted build system. This approach assumes that we'll be performing the authentication step for Goma at init-time, and then if we've stated we are using goma but are not authed properly an error is thrown at build-time. 

cc @MarshallOfSound @jkleinsc

### CODE REVIEW

File: src/util.js [hunk 1]
```diff
@@ -178,6 +197,11 @@ module.exports = {
   ensureDir,
   fatal,
   resolvePath,
+  authenticateGoma,
```
Comment (MarshallOfSound) [inline src/util.js]: We don't need this right?

File: src/e-build.js [hunk 2]
```diff
@@ -26,7 +26,15 @@ function ensureGNGen(config) {
 }
 
 function runNinja(config, target, ninjaArgs) {
-  sccache.ensure(config);
+  if (config.goma) {
+    const authenticated = goma.isAuthenticated(config.root);
+    if (!authenticated) {
+      throw new Error('Goma not authenticated.');
+    }
+  } else {
```
Comment (MarshallOfSound) [inline src/e-build.js]: We need a call here to `goma.ensure` that runs `goma_ctl.py ensure_start`

File: src/e-init.js [hunk 3]
```diff
@@ -16,15 +16,19 @@ function getVSReleaseLine() {
...
 
   // build the `gn gen` args
-  const gn_args = [
-    `import("//electron/build/args/${options.import}.gn")`,
-    `cc_wrapper="${sccache.exec(root)}"`,
-  ];
+  const gn_args = [`import("//electron/build/args/${options.import}.gn")`];
+
+  if (options.useGoma) {
+    gn_args.push('import("//electron/build/args/goma.gn")');
```
Comment (MarshallOfSound) [inline src/e-init.js]: We should only do this if this file actually exists.  `build-tools` still needs to work for < master

File: src/util.js [hunk 4]
```diff
@@ -105,8 +104,46 @@ function depotExecFileSync(config, exec, args, opts_in) {
   childProcess.execFileSync(exec, args, opts);
 }
 
+const getExternalBinaries = root => path.resolve(root, 'src', 'electron', 'external_binaries');
+
+function gomaIsAuthenticated(root) {
+  const gomaDir = path.resolve(getExternalBinaries(root), 'goma');
+
+  // Bail early if we're not on a branch with the ability to use Goma
+  if (fs.existsSync(gomaDir)) return;
```
Comment (MarshallOfSound) [inline src/util.js]:
```suggestion
  if (!fs.existsSync(gomaDir)) return;
```

File: src/util.js [hunk 5]
```diff
@@ -105,8 +104,46 @@ function depotExecFileSync(config, exec, args, opts_in) {
...
+
+  const loggedInPattern = /^Login as (\w+\s\w+)$/;
+  return loggedInPattern.test(loggedInInfo.toString().trim());
+}
+
+function authenticateGoma(root) {
+  const gomaDir = path.resolve(getExternalBinaries(root), 'goma');
+
+  // Bail early if we're not on a branch with the ability to use Goma
+  if (fs.existsSync(gomaDir)) return;
```
Comment (MarshallOfSound) [inline src/util.js]:
```suggestion
  if (!fs.existsSync(gomaDir)) return;
```

File: src/util.js [hunk 6]
```diff
@@ -105,8 +104,46 @@ function depotExecFileSync(config, exec, args, opts_in) {
...
+  if (!gomaIsAuthenticated(root)) {
+    childProcess.execFileSync('python', ['goma_auth.py', 'login'], { cwd: gomaDir });
+  }
+}
+
+function ensureGomaStart(root) {
+  const gomaDir = path.resolve(getExternalBinaries(root), 'goma');
+
+  // Bail early if we're not on a branch with the ability to use Goma
+  if (fs.existsSync(gomaDir)) return;
```
Comment (MarshallOfSound) [inline src/util.js]:
```suggestion
  if (!fs.existsSync(gomaDir)) return;
```

**Comment by codebytere** [comment]
hammering a few things out, hold for a few :)

**Pull request merged**